contrib/node/src/python/pants/contrib/node/tasks/apache_thrift_node_gen.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import json
+import os
+import re
+from typing import List
+
+from pants.backend.codegen.thrift.lib.apache_thrift_gen_base import ApacheThriftGenBase
+from pants.base.exceptions import TaskError
+from pants.build_graph.target import Target
+from pants.util.dirutil import safe_file_dump
+
+from pants.contrib.node.targets.node_module import NodeModule
+from pants.contrib.node.targets.node_thrift_library import NodeThriftLibrary
+
+
+class ApacheThriftNodeGen(ApacheThriftGenBase):
+  """Generate Nodejs source files from thrift IDL files"""
+  gentarget_type = NodeThriftLibrary
+  thrift_generator = 'js:node'
+  gen_directory = 'nodejs'
+
+  _COMPILER = 'thrift'
+
+  sources_globs = ('**/*',)
+
+  def synthetic_target_type(self, target: Target) -> NodeModule:
+    return NodeModule
+
+  @property
+  def _copy_target_attributes(self) -> List[str]:
+    return ['tags', 'node_scope', 'package_manager', 'build_script',
+            'output_dir', 'dev_dependency', 'style_ignore_path', 'bin_executables']
+
+  def execute_codegen(self, target, target_workdir):
+    sources_list = target.sources_relative_to_target_base()
+    if 'yarn.lock' not in sources_list.files:
```

Comment:
@nsaechao : Is the idea here that the thrift target will have an associated `yarn.lock` file? That seems fairly unlikely... does yarn actually use the `yarn.lock` file of a dependency, or only the one at the root where you called `install`?

Reply:
@stuhood: Yes, a `yarn.lock` file will be present. Currently we restrict our users from creating their own. This way we can control the `package.json` and the `yarn.lock` file. We create defaults for these which includes the `thrift` package dependency.